### PR TITLE
Fix error-index redirect to work with the back button.

### LINKED ISSUE #106485: Back button doesn't work from error_codes/error-index.html
Was looking thought the doc.rust-lang.org site to see what was new and looked at the compiler error codes page. when trying to navigate back to the main site by pressing back it simply flashes "if you are not redirected please ..." and then immediately goes back to error_codes/error-index.html. Pretty easy to repro, can provide more detail if needed!

### PR DESCRIPTION
**Pull request by ehuss**
This fixes the redirect page at https://doc.rust-lang.org/error-index.html so that it works with the browser's back-button. The solution is to use `window.location.replace()`, which avoids adding the page to the browser history stack.

This also cleans up the code a little bit, since it looks like it was written with different assumptions (maybe before f5857d5c5e1d2fde302f330d11c5cdea8005eb2a).  I don't think there is a need to have a redirect at https://doc.rust-lang.org/error_codes/error-index.html since I don't think fragment-based links were ever used there.

I have tested with Firefox, Chrome, and Safari. Going to `/error-index.html` redirects without adding to the stack, and can use the back button. Additionally, `/error-index.html#E0005` redirects correctly to the error page.

Finally, there is an unrelated commit to remove the 404 hack. There is an official way of avoiding the generation of the 404 page with setting input-404 to an empty value.

Fixes #106485

### CODE REVIEW

**Review by notriddle — APPROVED** [review]

File: src/tools/error_index_generator/redirect.js [hunk 1]
```diff
@@ -1,16 +0,0 @@
-(function() {
-    if (window.location.hash) {
-        let code = window.location.hash.replace(/^#/, '');
-        // We have to make sure this pattern matches to avoid inadvertently creating an
-        // open redirect.
```
[the remark this answers is not in the document]
Comment (ehuss) [reply src/tools/error_index_generator/redirect.js:5]: Oh, good point, thanks!

File: src/tools/error_index_generator/main.rs [hunk 2]
```diff
@@ -159,18 +160,27 @@ This page lists all the error codes emitted by the Rust compiler.
...
         <meta name="description" content="Book listing all Rust error codes">
-        <script src="error_codes/redirect.js"></script>
+        <script>
+            (function() {
+                if (window.location.hash) {
+                    let code = window.location.hash.replace(/^#/, '');
+                    // We have to make sure this pattern matches to avoid
+                    // inadvertently creating an open redirect.
+                    if (/^E[0-9]+$/.test(code)) {
+                        window.location.replace('./error_codes/' + code + '.html');
```
Comment (GuillaumeGomez) [inline src/tools/error_index_generator/main.rs:170]: Why not making this change in the JS file directly? It had the advantage of keeping things split (JS split from HTML) which is easier to read/maintain.

File: src/tools/error_index_generator/book_config.toml [hunk 3]
```diff
@@ -7,6 +7,7 @@ src = ""
 git-repository-url = "https://github.com/rust-lang/rust/"
 additional-css = ["error-index.css"]
 additional-js = ["error-index.js"]
+input-404 = ""
```
Comment (GuillaumeGomez) [inline src/tools/error_index_generator/book_config.toml:10]: Nice improvement!

File: src/tools/error_index_generator/main.rs (hunk 2 above)
Comment (ehuss) [reply src/tools/error_index_generator/main.rs]: Oh, no particular reason. It was no longer being shared between different pages, and it seemed a little more convenient to have everything in one place. I don't have a strong preference, though, so I pushed an update to keep it in the old location.

Comment (GuillaumeGomez) [reply src/tools/error_index_generator/main.rs]: Thanks! It also makes the changes much smaller. :)

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me. Thanks for the improvement!

**Pull request merged**